### PR TITLE
Refactor and improve join benchmarks with nvbench

### PR DESCRIPTION
**Pull request by PointKernel**
This PR includes several changes related to join benchmarks:

- Porting `inner_join` benchmarks in `JOIN_BENCH` to nvbench
- Adding benchmarks for `left_join` and `full_join`
- Creating a new header file (`join_benchmark_common.hpp`) to incorporate common APIs for join benchmarks
- Removing redundant code

### CODE REVIEW

**Comment by harrism** [comment]
@PointKernel this PR was made ready for review after burndown started. I don't think it's essential for 21.08 and it doesn't have reviews yet, so lets move to 21.10.  Thanks!

**Comment by PointKernel** [comment]
rerun tests

File: cpp/benchmarks/join/join_benchmark_common.hpp [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    }
+  }();
+  auto const probe_table_size = [&]() {
+    if constexpr (std::is_same<state_type, benchmark::State>::value) {
+      return static_cast<cudf::size_type>(state.range(1));
+    }
+    if constexpr (std::is_same<state_type, nvbench::state>::value) {
+      return static_cast<cudf::size_type>(state.get_int64("Probe Table Size"));
+    }
+  }();
```
Comment (karthikeyann) [inline cpp/benchmarks/join/join_benchmark_common.hpp:62]: default return missing. Is this not throwing a warning during compilation?

File: cpp/benchmarks/join/join_benchmark_common.hpp [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+  auto build_random_null_mask = [&rand_gen](int size) {
+    if (Nullable) {
+      // roughly 25% nulls
+      auto validity = thrust::make_transform_iterator(
+        thrust::make_counting_iterator(0),
+        [&rand_gen](auto i) { return (rand_gen.generate() & 3) == 0; });
+      return cudf::test::detail::make_null_mask(validity, validity + size);
+    } else {
+      return cudf::create_null_mask(size, cudf::mask_state::UNINITIALIZED);
+    }
```
Comment (karthikeyann) [inline cpp/benchmarks/join/join_benchmark_common.hpp:79]: is this needed?
`build_random_null_mask` is used when `Nullable` is true.
```suggestion
    }
```

File: cpp/benchmarks/join/join_benchmark_common.hpp [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+  CHECK_CUDA(0);
+
+  cudf::table_view build_table({build_key_column->view(), build_payload_column});
+  cudf::table_view probe_table({probe_key_column->view(), probe_payload_column});
+
+  // Setup join parameters and result table
+  [[maybe_unused]] std::vector<cudf::size_type> columns_to_join = {0};
+
+  // Benchmark the inner join operation
+  if constexpr (std::is_same<state_type, benchmark::State>::value and (not is_conditional)) {
```
Comment (karthikeyann) [inline cpp/benchmarks/join/join_benchmark_common.hpp:122]:
```suggestion
  if constexpr (std::is_same_v<state_type, benchmark::State> and (not is_conditional)) {
```
Also at other places.

File: cpp/benchmarks/join/join_benchmark_common.hpp [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+#pragma once
+
+#include <benchmark/benchmark.h>
+#include <nvbench/nvbench.cuh>
+
+#include <thrust/iterator/counting_iterator.h>
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/join.hpp>
+#include <cudf/table/table.hpp>
```
Comment (karthikeyann) [inline cpp/benchmarks/join/join_benchmark_common.hpp:26]: unused include?

**Review by karthikeyann — CHANGES_REQUESTED** [review]

File: cpp/benchmarks/join/join_benchmark_common.hpp (hunk 1 above)
Comment (PointKernel) [reply cpp/benchmarks/join/join_benchmark_common.hpp:62]: The two `if constexpr`s are to get rid of the warning you mentioned where NVCC will warn about a missing return when using an if constexpr/else.

File: cpp/benchmarks/join/join_benchmark_common.hpp (hunk 2 above)
Comment (PointKernel) [reply cpp/benchmarks/join/join_benchmark_common.hpp:79]: Good point! Thanks!

**Comment by PointKernel** [comment]
rerun tests

**Review by karthikeyann — APPROVED** [review]

**Comment by PointKernel** [comment]
rerun tests

**Comment by PointKernel** [comment]
@gpucibot merge

**Pull request merged**